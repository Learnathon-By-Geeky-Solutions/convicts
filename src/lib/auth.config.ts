import type { NextAuthConfig, Session } from "next-auth"
import { JWT } from "next-auth/jwt"
import GitHub from "next-auth/providers/github"
import Google from "next-auth/providers/google"

interface JWTSessionParams {
  token: JWT
  session: Session
}

export default {
  pages: { signIn: "/login" },
  session: { strategy: "jwt" },
  callbacks: {
    async authorized({ auth, request: { nextUrl } }) {
      const isSVG = nextUrl.pathname.endsWith(".svg")

      // newUser is only relevant for email sign-ins
      const oldUserLoggedIn = !!auth && !auth.newUser

      if (nextUrl.pathname.startsWith("/logout")) return true
      if (nextUrl.pathname.startsWith("/home")) return oldUserLoggedIn

      if (auth && !isSVG)
        return oldUserLoggedIn
          ? Response.redirect(new URL("/home", nextUrl))
          : false

      return true
    },

    async session({ token, session }: JWTSessionParams) {
      session.newUser = token.newUser
      if (session.user) {
        session.user.username = token.username
        session.user.name = token.name
      }
      return session
    },
  },
  providers: [Google, GitHub],
} satisfies NextAuthConfig
