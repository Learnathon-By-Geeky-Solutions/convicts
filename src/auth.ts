import authConfig from "@/auth.config"
import { PrismaAdapter } from "@auth/prisma-adapter"
import NextAuth from "next-auth"
import NodeMailer from "next-auth/providers/nodemailer"

import { prisma } from "@/lib/prisma"

export const { handlers, signIn, signOut, auth, unstable_update } = NextAuth({
  ...authConfig,
  adapter: PrismaAdapter(prisma),
  providers: [
    ...authConfig.providers,
    NodeMailer({
      server: {
        host: process.env.EMAIL_SERVER_HOST,
        port: parseInt(process.env.EMAIL_SERVER_PORT!),
        auth: {
          type: "OAuth2",
          user: process.env.EMAIL_SERVER_USER,
          clientId: process.env.EMAIL_CLIENT_ID,
          clientSecret: process.env.EMAIL_CLIENT_SECRET,
          refreshToken: process.env.EMAIL_REFRESH_TOKEN,
          accessToken: process.env.EMAIL_ACCESS_TOKEN,
        },
      },
      from: process.env.EMAIL_FROM,
    }),
  ],

  callbacks: {
    ...authConfig.callbacks,
    async jwt({ token, trigger, account, session }) {
      if (trigger === "signUp" && account!.type === "email") {
        const email = account!.providerAccountId
        const user = await prisma.user.findUnique({ where: { email } })
        token.newUser = user?.username == null
      } else if (trigger === "update") token.newUser = session.newUser

      return token
    },
  },
})
