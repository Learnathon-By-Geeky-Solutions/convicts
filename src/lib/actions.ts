"use server"

import { auth, unstable_update as updateSession } from "@/auth"
import { redirect } from "next/navigation"
import { z } from "zod"

import { prisma } from "@/lib/prisma"
import { UserSchema } from "@/lib/utils"

const FirstLoginSchema = UserSchema.pick({
  username: true,
  name: true,
  email: true,
})

export async function firstLogin(
  prevState: z.ZodError<z.infer<typeof FirstLoginSchema>> | void,
  formData: FormData,
) {
  const session = await auth()
  const data: { [key: string]: string } = {}
  formData.forEach((value, key) => (data[key] = value.toString()))
  data.email = session!.user!.email!

  const parsedInput = await FirstLoginSchema.safeParseAsync(data)
  if (!parsedInput.success) return parsedInput.error

  const { username, name, email } = parsedInput.data
  await prisma.user.update({
    where: { email: email },
    data: {
      username,
      name,
    },
  })

  const user = { ...session!.user, username, name }
  await updateSession({ ...session, user, newUser: false })
  redirect("/home")
}
