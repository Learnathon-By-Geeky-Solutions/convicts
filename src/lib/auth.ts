import React from "react"

import { PrismaAdapter } from "@auth/prisma-adapter"
import { render } from "@react-email/components"
import NextAuth from "next-auth"
import NodeMailer from "next-auth/providers/nodemailer"
import nodemailer from "nodemailer"

import VerifyEmail from "@/emails/verification"

import authConfig from "@/lib/auth.config"
import { prisma } from "@/lib/prisma"

// import { createId } from "@paralleldrive/cuid2"

export const { handlers, signIn, signOut, auth, unstable_update } = NextAuth({
  ...authConfig,
  adapter: PrismaAdapter(prisma),

  // in case username must be non null
  // this is a nice implementation specific workaround

  // adapter: PrismaAdapter(
  //   prisma.$extends({
  //     query: {
  //       user: {
  //         async create({ args, query }) {
  //           args.data.username ??= `!${createId()}`
  //           return query(args)
  //         },
  //       },
  //     },
  //   }),
  // ),

  providers: [
    ...authConfig.providers,
    NodeMailer({
      server: {
        host: process.env.EMAIL_SERVER_HOST,
        port: Number.parseInt(process.env.EMAIL_SERVER_PORT!),
        auth: {
          type: "OAuth2",
          user: process.env.EMAIL_SERVER_USER,
          clientId: process.env.EMAIL_CLIENT_ID,
          clientSecret: process.env.EMAIL_CLIENT_SECRET,
          refreshToken: process.env.EMAIL_REFRESH_TOKEN,
          accessToken: process.env.EMAIL_ACCESS_TOKEN,
        },
      },
      async sendVerificationRequest({ provider, url, identifier: to }) {
        const transporter = nodemailer.createTransport(provider.server)
        const subject = "Sign in to Carlander"
        const email = React.createElement(VerifyEmail, { url })

        const [html, text] = await Promise.all([
          render(email),
          render(email, { plainText: true }),
        ])

        await transporter.sendMail({ to, html, text, subject })
      },
    }),
  ],

  callbacks: {
    ...authConfig.callbacks,
    async jwt({ token, trigger, session }) {
      if (trigger === "signUp") token.newUser = true
      else if (trigger === "update") {
        token.newUser = session.newUser
        token.username = session.user.username
        token.name = session.user.name
      }

      return token
    },
  },
})
