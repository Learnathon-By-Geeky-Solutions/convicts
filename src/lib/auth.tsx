import { PrismaAdapter } from "@auth/prisma-adapter"
import { render } from "@react-email/components"
import NextAuth from "next-auth"
import NodeMailer from "next-auth/providers/nodemailer"
import nodemailer from "nodemailer"

import VerifyEmail from "@/emails/verification"

import authConfig from "@/lib/auth.config"
import { prisma } from "@/lib/prisma"

export const { handlers, signIn, signOut, auth, unstable_update } = NextAuth({
  ...authConfig,
  adapter: PrismaAdapter(prisma),
  providers: [
    ...authConfig.providers,
    NodeMailer({
      server: {
        host: process.env.EMAIL_SERVER_HOST,
        port: Number.parseInt(process.env.EMAIL_SERVER_PORT!),
        auth: {
          type: "OAuth2",
          user: process.env.EMAIL_SERVER_USER,
          clientId: process.env.EMAIL_CLIENT_ID,
          clientSecret: process.env.EMAIL_CLIENT_SECRET,
          refreshToken: process.env.EMAIL_REFRESH_TOKEN,
          accessToken: process.env.EMAIL_ACCESS_TOKEN,
        },
      },
      async sendVerificationRequest({ provider, url, identifier: to }) {
        const transporter = nodemailer.createTransport(provider.server)
        const subject = "Sign in to Carlander"

        const [html, text] = await Promise.all([
          render(<VerifyEmail url={url} />),
          render(<VerifyEmail url={url} />, { plainText: true }),
        ])

        await transporter.sendMail({ to, html, text, subject })
      },
    }),
  ],

  callbacks: {
    ...authConfig.callbacks,
    async jwt({ token, trigger, account, session }) {
      if (trigger === "signUp" && account!.type === "email") {
        const email = account!.providerAccountId
        const user = await prisma.user.findUnique({ where: { email } })
        token.newUser = user?.username == null
      } else if (trigger === "update") {
        token.newUser = session.newUser
        token.username = session.user.username
        token.name = session.user.name
      }

      return token
    },
  },
})
