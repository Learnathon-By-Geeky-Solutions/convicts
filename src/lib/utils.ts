import { Prisma } from "@prisma/client"
import { clsx, type ClassValue } from "clsx"
import { twMerge } from "tailwind-merge"
import { z } from "zod"

import { prisma } from "@/lib/prisma"

import UserUncheckedCreateInput = Prisma.UserUncheckedCreateInput

export const UserSchema = z.object({
  id: z.string().cuid().optional(),
  username: z
    .string()
    .min(3)
    .max(20)
    .regex(/(?=.*[a-zA-Z])[\w-]*/, {
      message: "Buckoo, stick to a reasonable username",
    })
    .refine(
      // check for unique username
      // prettier-ignore
      async (username) => !(await prisma.user.findUnique({ where: { username } })),
      { message: "Username is already taken" },
    ),
  name: z.string().min(3).max(64).optional(),
  email: z.string().email(),
  image: z.string().optional(),
  createdAt: z.string().datetime().optional(),
  updatedAt: z.string().datetime().optional(),
}) satisfies z.Schema<UserUncheckedCreateInput>

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs))
}
