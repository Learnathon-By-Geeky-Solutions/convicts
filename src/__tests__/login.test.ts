import { render, screen } from "@testing-library/react"

import LoginPage from "@/app/login/page"

import { auth } from "@/lib/auth"

jest.mock("@/lib/auth", () => ({
  auth: jest.fn(),
  signIn: jest.fn(),
}))

describe("LoginPage", () => {
  it("renders login page unchanged", async () => {
    const { container } = render(await LoginPage())
    expect(container).toMatchSnapshot()
  })

  it("renders the login page for new users", async () => {
    ;(auth as jest.Mock).mockResolvedValue({
      newUser: true,
      user: { name: "Tester" },
    })

    render(await LoginPage())

    expect(
      screen.getByText("Quickstart with the following details"),
    ).toBeInTheDocument()
    expect(screen.getByText("Carlander")).toBeInTheDocument()
    expect(screen.getByText("LOGO")).toBeInTheDocument()
  })

  it("renders the login page for existing users", async () => {
    ;(auth as jest.Mock).mockResolvedValue({ newUser: false })

    render(await LoginPage())

    expect(
      screen.getByText("Welcome! Please log in to continue"),
    ).toBeInTheDocument()
    expect(screen.getByText("Log in using")).toBeInTheDocument()
    expect(screen.getByText("Carlander")).toBeInTheDocument()
    expect(screen.getByText("LOGO")).toBeInTheDocument()
  })

  it("renders OAuthButtons for existing users", async () => {
    ;(auth as jest.Mock).mockResolvedValue({ newUser: false })

    render(await LoginPage())

    expect(screen.getByText("Log in using")).toBeInTheDocument()
    expect(screen.getByRole("button", { name: /Google/i })).toBeInTheDocument()
    expect(screen.getByRole("button", { name: /Github/i })).toBeInTheDocument()
  })

  it("renders FirstLogin component for new users", async () => {
    ;(auth as jest.Mock).mockResolvedValue({
      newUser: true,
      user: { name: "Tester" },
    })

    render(await LoginPage())

    expect(
      screen.getByText("Quickstart with the following details"),
    ).toBeInTheDocument()
    expect(screen.getByDisplayValue("Tester")).toBeInTheDocument()
  })
})
