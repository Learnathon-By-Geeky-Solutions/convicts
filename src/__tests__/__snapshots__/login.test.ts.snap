// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage renders login page unchanged 1`] = `
<div>
  <div
    class="mx-auto flex w-max max-w-sm flex-col items-center className"
  >
    <div
      class="flex flex-col items-center"
    >
      <span
        class="relative flex shrink-0 overflow-hidden mt-[4rem] aspect-square h-auto w-max rounded-lg border-4 border-zinc-950"
      >
        <span
          class="flex h-full w-full items-center justify-center bg-zinc-100 dark:bg-zinc-800 className rounded-none p-2 indent-[.2em] text-xl tracking-[.2em]"
        >
          LOGO
        </span>
      </span>
      <h1
        class="mt-10 mb-6 indent-[.3em] text-2xl tracking-[.3em] uppercase"
      >
        Carlander
      </h1>
      <p
        class="my-6 text-center"
      >
        Welcome! Please log in to continue
        <br />
        <span
          class="text-lime-500"
        >
          Log in using
        </span>
      </p>
      <form
        action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
        aria-label="oauth-buttons"
        class="my-9 flex gap-6"
      >
        <button
          class="inline-flex items-center justify-center gap-2 border-2 border-transparent text-sm font-medium whitespace-nowrap focus-visible:ring-1 focus-visible:ring-zinc-950 focus-visible:outline-hidden disabled:pointer-events-none disabled:opacity-50 dark:focus-visible:ring-zinc-300 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 bg-lime-500 text-zinc-50 shadow-xs transition-colors duration-100 hover:border-zinc-950 active:bg-zinc-950 h-auto rounded-xl px-[3.3rem] py-1.5"
          name="provider"
          value="google"
        >
          <img
            alt="Sign in with Google"
            data-nimg="1"
            decoding="async"
            height="36"
            loading="lazy"
            src="/google_logo.svg"
            style="color: transparent;"
            width="37"
          />
        </button>
        <button
          class="inline-flex items-center justify-center gap-2 border-2 border-transparent text-sm font-medium whitespace-nowrap focus-visible:ring-1 focus-visible:ring-zinc-950 focus-visible:outline-hidden disabled:pointer-events-none disabled:opacity-50 dark:focus-visible:ring-zinc-300 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 bg-lime-500 text-zinc-50 shadow-xs transition-colors duration-100 hover:border-zinc-950 active:bg-zinc-950 h-auto rounded-xl px-[3.3rem] py-1.5"
          name="provider"
          value="github"
        >
          <img
            alt="Sign in with Github"
            data-nimg="1"
            decoding="async"
            height="37"
            loading="lazy"
            src="/github_logo.svg"
            style="color: transparent;"
            width="38"
          />
        </button>
      </form>
    </div>
    <div
      class="flex w-xs items-center gap-6"
    >
      <div
        class="bg-zinc-200 dark:bg-zinc-800 h-[1px] w-full shrink border-2 border-lime-500"
        data-orientation="horizontal"
        role="none"
      />
      or
      <div
        class="bg-zinc-200 dark:bg-zinc-800 h-[1px] w-full shrink border-2 border-lime-500"
        data-orientation="horizontal"
        role="none"
      />
    </div>
    <form
      action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
      class="className mt-7 mb-11 flex min-w-xs flex-col gap-7"
    >
      <div>
        <label
          class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
          for="email"
        >
          Email Address
        </label>
        <input
          class="flex h-9 w-full bg-transparent px-3 py-1 text-base shadow-xs transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium file:text-zinc-950 placeholder:text-zinc-500 focus-visible:ring-1 focus-visible:outline-hidden disabled:cursor-not-allowed disabled:opacity-50 md:text-sm dark:border-zinc-800 dark:file:text-zinc-50 dark:placeholder:text-zinc-400 dark:focus-visible:ring-zinc-300 mt-2 rounded-lg border-3 border-lime-500 focus-visible:ring-lime-500"
          id="email"
          name="email"
          required=""
          type="email"
        />
      </div>
      <button
        class="inline-flex items-center justify-center gap-2 rounded-md border-2 border-transparent text-sm font-medium whitespace-nowrap focus-visible:ring-1 focus-visible:ring-zinc-950 focus-visible:outline-hidden disabled:pointer-events-none disabled:opacity-50 dark:focus-visible:ring-zinc-300 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 bg-zinc-950 text-zinc-50 shadow-xs transition-colors duration-100 hover:border-lime-500 active:bg-lime-500 h-9 px-4 py-2 uppercase"
        type="submit"
      >
        Continue
      </button>
    </form>
  </div>
</div>
`;
